### ISSUE TITLE
how to use focal loss in fast rcnn

### ISSUE THREAD

**Issue by wytcsuch** [facebookresearch/Detectron#763]
Hi，I want to apply focal loss to the fast RCNN part. That is to say, I still use RPN + FPN in front of fast rcnn, I just want to modify the loss of fast rcnn.So I change the code as follows:
In fast_rcnn_head.py:
`def add_fast_rcnn_losses(model):
    """Add losses for RoI classification and bounding box regression."""
    model.AddMetrics(['fg_num', 'bg_num'])
    loss_cls, cls_prob = model.net.SoftmaxFocalLoss(
        ['cls_score', 'labels_int32','fg_num'], ['loss_cls', 'cls_prob'],
        gamma=cfg.FOCAL_LOSS.LOSS_GAMMA,
        alpha=cfg.FOCAL_LOSS.LOSS_ALPHA,
        scale=model.GetLossScale(),
        num_classes=model.num_classes
    )`
and In fast_rcnn.py,I add:
`def get_fast_rcnn_blob_names(is_training=True):
    """Fast R-CNN blob names."""
    # rois blob: holds R regions of interest, each is a 5-tuple
    # (batch_idx, x1, y1, x2, y2) specifying an image batch index and a
    # rectangle (x1, y1, x2, y2)
    if is_training and cfg.FOCAL_LOSS.USE_FL:
        #fast rcnn+focal loss
        blob_names += ['fg_num']
        blob_names += ['bg_num']`

`def add_fast_rcnn_blobs(blobs, im_scales, roidb):
    """
    Add blobs needed for training Fast R-CNN style models.
    """
    # Sample training RoIs from each image and append them to the blob lists
    if cfg.FOCAL_LOSS.USE_FL:
        blobs['fg_num'], blobs['bg_num'] = 0.0, 0.0
        for im_i, entry in enumerate(roidb):
            frcn_blobs, fg_num, bg_num = _sample_rois_fl(entry, im_scales[im_i], im_i)
            for k, v in frcn_blobs.items():  
                blobs[k].append(v)
            blobs['fg_num'] += fg_num
            blobs['bg_num'] += bg_num
        blobs['fg_num'] = blobs['fg_num'].astype(np.float32)
        blobs['bg_num'] = blobs['bg_num'].astype(np.float32)
    else:
        for im_i, entry in enumerate(roidb):
            frcn_blobs = _sample_rois(entry, im_scales[im_i], im_i)
            for k, v in frcn_blobs.items():
                blobs[k].append(v)`

`def _sample_rois_fl(roidb, im_scale, batch_idx):
    """Generate a random sample of RoIs comprising foreground and background
    examples.
    """
    rois_per_image = int(cfg.TRAIN.BATCH_SIZE_PER_IM)  #128
    fg_rois_per_image = int(np.round(cfg.TRAIN.FG_FRACTION * rois_per_image)) 
    max_overlaps = roidb['max_overlaps']

    # Select foreground RoIs as those with >= FG_THRESH overlap
    fg_inds = np.where(max_overlaps >= cfg.TRAIN.FG_THRESH)[0]
    # Guard against the case when an image has fewer than fg_rois_per_image
    # foreground RoIs
    fg_rois_per_this_image = np.minimum(fg_rois_per_image, fg_inds.size)
    # Sample foreground regions without replacement
    if fg_inds.size > 0:
        fg_inds = npr.choice(
            fg_inds, size=fg_rois_per_this_image, replace=False
        )

    # Select background RoIs as those within [BG_THRESH_LO, BG_THRESH_HI)
    bg_inds = np.where(
        (max_overlaps < cfg.TRAIN.BG_THRESH_HI) &
        (max_overlaps >= cfg.TRAIN.BG_THRESH_LO)
    )[0]
    # Compute number of background RoIs to take from this image (guarding
    # against there being fewer than desired)
    bg_rois_per_this_image = rois_per_image - fg_rois_per_this_image
    bg_rois_per_this_image = np.minimum(bg_rois_per_this_image, bg_inds.size)
    # Sample foreground regions without replacement
    if bg_inds.size > 0:
        bg_inds = npr.choice(
            bg_inds, size=bg_rois_per_this_image, replace=False
        )

    # The indices that we're selecting (both fg and bg)
    keep_inds = np.append(fg_inds, bg_inds)
    # Label is the class each RoI has max overlap with 
    sampled_labels = roidb['max_classes'][keep_inds]
    sampled_labels[fg_rois_per_this_image:] = 0  # Label bg RoIs with class 0
    sampled_boxes = roidb['boxes'][keep_inds]

    bbox_targets, bbox_inside_weights = _expand_bbox_targets(
        roidb['bbox_targets'][keep_inds, :]
    )
    bbox_outside_weights = np.array(
        bbox_inside_weights > 0, dtype=bbox_inside_weights.dtype
    )
    #pdb.set_trace()
    # Scale rois and format as (batch_idx, x1, y1, x2, y2)
    sampled_rois = sampled_boxes * im_scale
    repeated_batch_idx = batch_idx * blob_utils.ones((sampled_rois.shape[0], 1))
    sampled_rois = np.hstack((repeated_batch_idx, sampled_rois))
    
    
    # Base Fast R-CNN blobs
    blob_dict = dict(
        labels_int32=sampled_labels.astype(np.int32, copy=False),
        rois=sampled_rois,
        bbox_targets=bbox_targets,
        bbox_inside_weights=bbox_inside_weights,
        bbox_outside_weights=bbox_outside_weights
    )

    # Optionally add Mask R-CNN blobs
    if cfg.MODEL.MASK_ON:
        mask_rcnn_roi_data.add_mask_rcnn_blobs(
            blob_dict, sampled_boxes, roidb, im_scale, batch_idx
        )

    # Optionally add Keypoint R-CNN blobs
    if cfg.MODEL.KEYPOINTS_ON:
        keypoint_rcnn_roi_data.add_keypoint_rcnn_blobs(
            blob_dict, roidb, fg_rois_per_image, fg_inds, im_scale, batch_idx
        )

    #pdb.set_trace()
    num_fg, num_bg = len(fg_inds), len(bg_inds)
    out_num_fg = np.array([num_fg + 1.0], dtype=np.float32)
    out_num_bg = (
            np.array([num_bg + 1.0]) * (cfg.MODEL.NUM_CLASSES - 1) +
            out_num_fg * (cfg.MODEL.NUM_CLASSES - 2))
            
    return blob_dict, out_num_fg, out_num_bg`

Then,I run train_net.py,It shows:
 [enforce fail at tensor_impl.h:781] i < dims_.size(). 2 vs 2. Exceeding ndim limitError from operator: 
input: "gpu_0/cls_score" input: "gpu_0/labels_int32" input: "gpu_0/fg_num" output: "gpu_0/loss_cls" output: "gpu_0/cls_prob" name: "" type: "SoftmaxFocalLoss" arg { name: "alpha" f: 0.25 } arg { name: "scale" f: 1 } arg { name: "num_classes" i: 3 } arg { name: "gamma" f: 2 } device_option { device_type: 1 cuda_gpu_id: 0 }
So,I turn to [https://github.com/pytorch/pytorch/blob/master/modules/detectron/softmax_focal_loss_op.cc](url),find that in softmax_focal_loss,the input data scores must be a 4D tensor with shape (N, C, H, W),but in fast rcnn,the shape of the scores is [batch_size, num_classes].
how to solve the problem?

**Comment by emmra** [comment]
The same problem to me, have you solved? thank you!!

**Comment by wytcsuch** [comment]
> The same problem to me, have you solved? thank you!!

sorry,I haven't solved it yet.

**Comment by ash1995** [comment]
@wytcsuch  Were you able to use focal loss with fast rcnn?
